extensions/spark/kyuubi-extension-spark-3-3/src/main/scala/org/apache/kyuubi/sql/RebalanceBeforeWriting.scala

```diff
@@ -27,7 +27,18 @@ trait RepartitionBuilderWithRebalance extends RepartitionBuilder {
       query: LogicalPlan): LogicalPlan = {
     if (!conf.getConf(KyuubiSQLConf.INFER_REBALANCE_AND_SORT_ORDERS) ||
       dynamicPartitionColumns.nonEmpty) {
-      RebalancePartitions(dynamicPartitionColumns, query)
+      // TODO: make configurable
+      val constantConditions = InferDynamicPartitionConstantConditions
+        .infer(dynamicPartitionColumns, query)
```

Comment:
We can add a new rule before `RepartitionBeforeWriting` that changes the dynamic partition write to static partition write.

Reply:
Are there any other benefits of changing to static partition writing? It may infer multiple constant partition conditions here.

Reply:
it saves a local sort at least..

Reply:
like: 

```
insert overwrite t1 partition(p1) select c1, p1 from t2 where p1 in (1', '2');
```

Reply:
For this case, does not `rebalance(t)` better than `rebalance` ? Can you talk more why `rebalance` is better ?

Reply:
> For this case, does not `rebalance(t)` better than `rebalance` ? Can you talk more why `rebalance` is better ?

Makes sense, I was eager to avoid large partitions and ignored the original intention of this optimization.

Reply:
> We can add a new rule before RepartitionBeforeWriting that changes the dynamic partition write to static partition write.

But this may be a dangerous behavior, if inference errors will lead to data errors, I think we can only control the behavior of Rebalance first, what do you think?

Reply:
Each optimization can cause bug, we should trust the code.. and we need a new config to control the behavior.

I think the key point is that, we should handle the multi-level partition spec carefully. e.g.,
```sql
INSERT INTO TABLE t1 (p1, p2, p3) SELECT * FROM t2 WHERE p1 = 'a'
```
For this case , it is still a dynamic partition write.

Reply:
> Each optimization can cause bug, we should trust the code.. and we need a new config to control the behavior.

Thanks, i will try to implement it.

> I think the key point is that, we should handle the multi-level partition spec carefully. e.g.,

The following judgment should cover it, I will add more test cases to verify it.

https://github.com/apache/kyuubi/blob/d1c9f4d935c6e32366eac019b955f72731ecebc4/extensions/spark/kyuubi-extension-spark-3-3/src/main/scala/org/apache/kyuubi/sql/RebalanceBeforeWriting.scala#L34-L35